### PR TITLE
Add lambda to handle public API request tracking

### PR DESCRIPTION
**Pull request by steventux**
https://trello.com/c/nbpTiCbi/1102-publish-aws-lambda-to-process-public-api-logs
also see https://trello.com/c/c0odHIAh/1101-get-fastly-to-send-cdn-logs-for-public-api-requests-to-s3

This lambda is designed to be triggered by ObjectCreated events on an S3 bucket receiving logs for public API requests.
Log files matching the expected bucket path will be read and GA tracking events created from their contents.
As advised, IP addresses are not currently tracked.

### CODE REVIEW

File: send_public_api_events_to_ga/send_public_api_events_to_ga.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        timestamp = timestamp or int(round(datetime.now().timestamp()))
+        params = urllib.parse.urlencode({
+	    'v': 1,
+	    'tid': 'UA-26179049-14',
+	    'cid': ga_client_id or 'No client ID present',
+	    't': 'event',
+	    'ec': 'Public API request',
+	    'ea': path or 'No path present',
+	    'el': referrer or 'No referrer present',
+	    'ua': user_agent or 'No user agent present',
```
Comment (kevindew) [inline send_public_api_events_to_ga/send_public_api_events_to_ga.py]: For these things when we have an `or` might it be better to just not send them? Feels like we pollute the store with these incorrect values.

File: send_public_api_events_to_ga/send_public_api_events_to_ga.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+def send_events_to_GA(s3_object):
+    urls = []
+    csvreader = csv.reader(open_for_read(s3_object), delimiter="\t")
+    for row in csvreader:
+        timestamp, status, path, ip, referrer, user_agent, ga_client_id = row
+        timestamp = timestamp or int(round(datetime.now().timestamp()))
+        params = urllib.parse.urlencode({
+	    'v': 1,
+	    'tid': 'UA-26179049-14',
```
[the remark this answers is not in the document]
Comment (steventux) [reply send_public_api_events_to_ga/send_public_api_events_to_ga.py]: I think it should, however configuring that requires someone with the appropriate access.

[the remark this answers is not in the document]
Comment (kevindew) [reply send_public_api_events_to_ga/send_public_api_events_to_ga.py]: Sure, well if there's a precedent set for hardcoding it and it's easier follow that. Might be prudent to add a comment though.

File: send_public_api_events_to_ga/send_public_api_events_to_ga.py (hunk 1 above)
Comment (steventux) [reply send_public_api_events_to_ga/send_public_api_events_to_ga.py]: Looking at the GA measurements protocol hit builder it's valid to not send param values for these so I'll remove the `or` values.

**Comment by kevindew** [comment]
Thanks for the fixes and the readme 👍 

I was able to run the `deploy.sh` but am still stuck running the tests.

With `deploy.sh` it might be worth stating that if you're using pyenv you need to install virtualenv by it (I got caught out by it using my system one)

Guess steps are:
```
pyenv install 3.6.0
pyenv local 3.6.0
pip install virtualenv
```
Might also be worth including the `.python_version` file for pyenv? I'm not sure though how many of these problems would be obvious to someone less of a python novice.

With the tests I get package not found for `nosetest` but was able to install `nose` which provided nosetests. However I still get a problem from lack of `mock` module.

```
nosetests ./test_send_public_api_events_to_ga.py
E
======================================================================
ERROR: Failure: ModuleNotFoundError (No module named 'mock')
```

Do we need to set up a virtualenv for the tests to run and have mock in the `requirements.txt` ? 

Happy to resolve these issues after deploying if my pains become a blocker.

**Comment by steventux** [comment]
@kevindew I've added `test_requirements.txt` excluded it from the deployment artefact, updated the README on how to install test dependencies and added the `.python-version` file to VC.

**Comment by kevindew** [comment]
@steventux great thanks - I just had a successful test run 🎉 

Just one tiny issue https://github.com/alphagov/govuk-lambda-app-deployment/pull/10/files#diff-396192fc29367e25391a436492afb9c3R38 the binary is nosetests rather than nosetest

**Comment by steventux** [comment]
Thanks, good spot, updated.

**Pull request merged**